etna/pipeline/hierarchical_pipeline.py

```diff
@@ -0,0 +1,117 @@
+from typing import Dict
+from typing import List
+from typing import Sequence
+
+from etna.datasets.tsdataset import TSDataset
+from etna.metrics import Metric
+from etna.models.base import ModelType
+from etna.pipeline.pipeline import Pipeline
+from etna.reconciliation.base import BaseReconciliator
+from etna.transforms.base import Transform
+
+
+class HierarchicalPipeline(Pipeline):
+    """Pipeline of transforms with a final estimator for hierarchical time series data."""
+
+    def __init__(
+        self, reconciliator: BaseReconciliator, model: ModelType, transforms: Sequence[Transform] = (), horizon: int = 1
+    ):
+        """Create instance of HierarchicalPipeline with given parameters.
+
+        Parameters
+        ----------
+        reconciliator:
+            Instance of reconciliation method
+        model:
+            Instance of the etna Model
+        transforms:
+            Sequence of the transforms
+        horizon:
+             Number of timestamps in the future for forecasting
+        """
+        super().__init__(model=model, transforms=transforms, horizon=horizon)
+        self.reconciliator = reconciliator
+
+    def fit(self, ts: TSDataset) -> "HierarchicalPipeline":
+        """Fit the Pipeline.
+
+        Fit and apply given transforms to the data, then fit the model on the transformed data.
+
+        Parameters
+        ----------
+        ts:
+            Dataset with timeseries data
+
+        Returns
+        -------
+        :
+            Fitted Pipeline instance
+        """
+        self.reconciliator.fit(ts=ts)
+        ts = self.reconciliator.aggregate(ts=ts)
+        super().fit(ts=ts)
+        return self
+
+    def raw_forecast(
+        self, prediction_interval: bool = False, quantiles: Sequence[float] = (0.25, 0.75), n_folds: int = 3
+    ) -> TSDataset:
+        """Make a prediction for target at the source level of hierarchy.
+
+        Parameters
+        ----------
+        prediction_interval:
+            If True returns prediction interval for forecast
+        quantiles:
+            Levels of prediction distribution. By default 2.5% and 97.5% taken to form a 95% prediction interval
+        n_folds:
+            Number of folds to use in the backtest for prediction interval estimation
+
+        Returns
+        -------
+        :
+            Dataset with predictions at the source level
+        """
+        forecast = super().forecast(prediction_interval=prediction_interval, quantiles=quantiles, n_folds=n_folds)
+        hierarchical_forecast = TSDataset(
```

Comment:
It is actually a tricky place, as we only need to add hierarchical structure to the dataset. Tis might cause the columns duplication after  `_merge_exog` operation and calling the constructor is quite expensive operation in fact. We need to come up with some work around

Reply:
Right now we don't have any other methods to convert dataset to hierarchical. Duplication can be solved by filtering out columns from `df` that appear in `df_exog` before calling constructor.

Reply:
As a workaround for now we can do something like this `hierarchical_forecast = TSDataset(df=forecast[..., "target"], ...)`

Reply:
We will do sth with this later